### ISSUE TITLE
How should we normalize the audio signal for wav2vec 2.0 inference?

### ISSUE THREAD

**Issue by TParcollet** [pytorch/fairseq#3277]
Thanks Fairseq and @alexeib for releasing wav2vec 2.0.

We currently are training multiple wav2vec 2.0 models based on the given configuration files. First, I noticed that only the large configuration includes the `normalisation: True` YAML parameter. Hence, my first question is: why not both? Did you see any bad behaviour by normalizing also for the base model? 

Then, I did a features extraction script by adapting the wav2vec 1 one. Basically, we are just calling the .forward() function. However, I suppose that we should normalize the audio signal in the same manner as it was done during training. Indeed, I did not see any normalization happening in the wav2vec `.forward()` function. Therefore, how should we define the normalisation so it exactly matches the one that is applied when using `normalization: True` with `fairseq-hydra-train` ? 

Thanks you so much,

**Comment by TParcollet** [comment]
I checked the code, and I can't find any place where the `normalize` argument is used appart from `audio_utils.py`. However, no normalisation happens in there. @HenryZhou7 @alexeib I believe that this is a very important point to address so peoples can build extraction scripts in top of your amazing work :)

**Comment by alexeib** [comment]
Hello,
The normalization is different because originally when developing wav2vec we relied on group norm in the first layer of the feature extractor to do the normalization for us, which worked well. When scaling the model to a larger size later on, we noticed that the model was unstable when trained over long periods of time, and one of the reasons was this group norm based normalization. We achieved stability by replacing group norm with layer norm in all feature extractor layers and moved input normalization into the data loader.

In general, for new models, you should probably train them with normalize=True and model.extractor_mode='layer_norm' even for smaller models. I kept the configs in the examples dir to be consistent with the published models that we describe in our paper, but I will update the docs when I get to them with new recommendations.

When you are evaluating models, you should use the same normalization scheme that was used for pre-training, otherwise you will notice significant accuracy drops.

To normalize yourself in the same way as "normalize: True" take a look here: https://github.com/pytorch/fairseq/blob/master/fairseq/data/audio/raw_audio_dataset.py#L60

**Comment by TParcollet** [comment]
Thank you very much for you detailed answer @alexeib.

We used the pytorch layer norm before feeding the pretrained model and it works well!
